Fix parsing of subscript expressions using python resolver

wconstab opened the pull request:
- add call out to python resolver in parseArgsFromDecl, parserReturnFromDecl
- add support in python resolver for nested subexpressions
- wrap python resolver call in exception handling to fall back to c++ path
- add tests for newly resolvable types 
- closes #38728

Fixes bug where SourceRange objects did not include the final closing ']' for a subscript expression.  E.g. range for 'List[int]' previously included only 'List[int'.

wconstab committed:
Fix parsing of subscript expressions

Context: e.g. 'torch.jit.Future[int]'
- parseSubscript called with value='torch.jit.Future'
- range is set to the first '[' token
- subscript_exprs includes the 'int' part
- the final ']' was dropped

This change creates a new SourceRange that spans from the '['
to the ']'.

wconstab committed:
hack to ensure closing ] is captured

wconstab committed:
force calling resolver in parseTypeFromExpr

wconstab committed:
parse square bracket subexpr in resolution callback

wconstab committed:
generalize the string passed to resolver_

- avoid throwing during resolver failure,
  fall back to orignal behavior

wconstab committed:
add tests and implement tuple and dict parsing

 - newly added test pass and verify the resolution callback
   changes work as intended
 - other existing test failures have been introduced and
   need to be debugged

wconstab committed:
reminder for additional test needed

wconstab committed:
fix tests in test_jit_py3.py

wconstab committed:
fix additonal tests

 - in cases where base_mod isn't resolved, bail out and fall back
 - handle more complex nesting of [ and ,
   e.g. List[Tuple[int, int]] don't let the inner ',' jump the gun,
        parse the Tuple subexpr first

wconstab commented on torch/_jit_internal.py:
```diff
@@ -31,12 +31,50 @@ def env(qualified_name, module):
+                parts = no_whitespace.split(',')
+                types = [env(p, module) for p in parts]
+
+                # HACK for now - must be a better way to approximate base_mod[*types]
```
@eellison Any suggestion for a better way to do this part?

wconstab commented on torch/_jit_internal.py:
```diff
@@ -31,12 +31,50 @@ def env(qualified_name, module):
             remainding_pieces = '.'.join(parts[1:])
             module_value = getattr(module, base)
             return env(remainding_pieces, module_value)
+        elif 'int' == qualified_name:
```
thoughts on doing this here? or should these type of things result in a bailout back to cpp too?

if this is the right approach, then i'd probably want to be more thorough about the types we handle.

wconstab replied:
i'll come back to this one; i only added it to fix some failures in running test_jit.py, and in those cases getattr(module, 'int') was failing for me i thought.

wconstab replied:
yea these were able to go away.  i had added them in to good effect, they made a test pass. but taking them out and adding in more thorough 'fallback' (return None) coverage fixed it too.

wconstab commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -162,13 +162,22 @@ c10::optional<std::string> ScriptTypeParser::parseBaseTypeName(
 TypePtr ScriptTypeParser::parseTypeFromExpr(const Expr& expr) const {
   if (expr.kind() == TK_SUBSCRIPT) {
+
+    if (resolver_) {
```
I put this in as a hack at first with a hardcoded string passed in.  using expr.range().text() seems to be correct in most cases. I'm not sure how to decide if it's actually an OK approach or still a hack.

eellison reviewed (commented):
this is looking really good ! A few small fixes we need to make but I think this is close. 

also mind updating the title of the PR and adding some context to it (maybe link to original issue)

wconstab committed:
handle tuple of arbitrary length

wconstab committed:
subexpression test improvements

- implement test for Optional type
- add type-specific usages inside functions to exercise the JIT

wconstab commented on test/test_jit_py3.py:
```diff
@@ -403,6 +403,40 @@ def wrong_type():
+
+    def test_unimported_type_resolution(self):
+        # verify fallback from the python resolver to the c++ resolver
+        assert False, "TODO implement me, maybe in a new file?"
```
oh, yea that's a good workaround; and I guess I can first assert in the test that the type isn't resolved, to guard against someone else importing it in the future.  I was overthinking it perhaps but wanted to handle things that already were imported, which i don't think i can do in the same file but maybe can do in another file...

wconstab replied:
ok i sorta made this test but i left it WIP.  Is the point just to test that it runs?  or is it supposed to invoke my code? i was also surprised that 'number' got implicitly converted to tensor, so i was hesitant to assert for Tensor in the FileCheck part.

wconstab committed:
annotate return types for subexpression tests, use FileCheck

wconstab commented on test/test_jit_py3.py:
```diff
@@ -403,6 +403,40 @@ def wrong_type():
+            return x
+
+        g = fn.graph
+        assert "Future" in repr(g)
```
cool. maybe for the optional case i can use check-dag? not sure how to best use it in that case.

wconstab committed:
call python resolver only at top level args/return parsing

wconstab committed:
fix spelling

wconstab committed:
fix handling of nested [] and , subexp

wconstab committed:
fixes for unit tests that expect assertions from c++ side

wconstab committed:
roll back what seems to be a bad fix

wconstab committed:
refactor subexpression parser

wconstab committed:
fix Tensor[Tensor] case

- e.g. fix the test that expects a particular assertion
- make sure this case results in bailing out back to c++ parser

wconstab committed:
WIP test for non-imported type

wconstab commented on torch/_jit_internal.py:
```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
+            subexp = "".join(subexp.split())
+            splits = []
+            nest_level = 0
+            for i, c in enumerate(subexp):
```
thoughts on this, cleaner/easier way to do it?

wconstab commented on torch/_jit_internal.py:
```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
+            base_mod = env(base, module)
+
+            # intentional bailout, e.g. base == "Tuple" but wasn't imported in Module
+            if not base_mod or isinstance(base_mod, torch.Tensor):
```
oops, the part after 'or' was meant to be cleaned up. removing it after retesting..

wconstab commented on torch/_jit_internal.py:
```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
     You should not use this directly, it should only be used from the other
     createResolutionCallbackFrom* functions.
     """
+    def try_build_type(base_mod, subexp_mod):
```
i started having logic in multiple places and it was still failing in cases like a unit test that tries Tensor[Tensor] and expects a particular exception.

wconstab committed:
cleanup

eellison reviewed (commented):
The current parsing logic is kind of complicated / hard to follow. I'm not sure exactly if if it works robustly (although the test cases look good). I have one suggestion of how to make it easier but looks almost ready to land.

wconstab commented on torch/_jit_internal.py:
```diff
@@ -26,17 +26,76 @@ def createResolutionCallbackFromEnv(lookup_base):
+                return try_build_type(base_mod, subexp_mod)
+        else:
+            # either base or subexp could include '.' or '['
+            if 'Tensor' == subexp:
```
yea, i think that's what this was from.  i'd only point out that i think it was failing in a way that was not gracefully falling back to the old parser, so i had to do something about it (either make it work, or make it return None).  I've leaned more toward the latter in general, but i'll have to revisit this case and see if i can just delete it now that I have checks/try/except in more places or if i'll still need some specific handling for it.

wconstab commented on test/test_jit_py3.py:
```diff
@@ -403,6 +403,51 @@ def wrong_type():
+            # type (number) -> number
+            return x + 1
+
+        FileCheck().check('number').run(fn.graph)
```
@eellison Clarification on what you were wanting this test to cover and what the expected behavior of the system is (i observed that the number gets converted to a Tensor which initially surprised me but may be expected - should I assert on Tensor?)

eellison replied:
numbers gets converted to a tensor with `x + 1` ? that is weird.

wconstab committed:
fix test_unimported_type_resolution

wconstab committed:
WIP refactor

wconstab committed:
WIP refactor almost working

wconstab committed:
WIP fixed py3 tests

wconstab committed:
passing tests

wconstab committed:
cleanup

wconstab committed:
assert whole expression is parsed in python parser

- the whole expression should have been parsed,
  if not, either raise the exception or fall back to
  c++ parser

eellison reviewed (approved):
This is awesome! Definitely a tricky PR, but is much needed to release `torch.jit.fork`. This will also make it super easy to handle https://docs.python.org/3.9/whatsnew/3.9.html#pep-585-builtin-generic-types without adding weird special handling to our script type parser. 

I think this also fixes: https://github.com/pytorch/pytorch/issues/16492 and https://github.com/pytorch/pytorch/issues/32544 and

wconstab commented on torch/_jit_internal.py:
```diff
@@ -26,21 +26,48 @@ def createResolutionCallbackFromEnv(lookup_base):
+            i += 1
+            part, part_len = parseNestedExpr(expr[i:], module)
+            parts.append(part)
+            i += part_len
```
yea. ugh. this re-breaks my fix in the c++ side parser i think. taking a look..

wconstab committed:
comment the expectations for try/except
